src/Runtime/Eval.hs

```diff
@@ -165,7 +165,7 @@ eval (Ref n) = do
                   -- valid reference
                   (Just v) -> case v of
                     -- Pending Value, need to eval this to get the actual value
-                    (Pv _ e') -> evalWithLimit $ eval e'
+                    (Pv env e') -> extScope env $ evalWithLimit $ eval e'
```

Comment:
Finally coming back around to this, and nice! This will work nicely, and I'll add in those other outstanding items that were mentioned some weeks ago.